### ISSUE TITLE
Compiling the vmbuilder fails

### ISSUE THREAD

**Issue by JMendyk** [hioa-cs/IncludeOS#942]
When I try to install IncludeOS using `./install.sh` compilation of vmbuilder fails with this error messages:
```
>>> Compiling the vmbuilder, which makes a bootable vm out of your service
~/.installs/IncludeOS/vmbuild ~/.installs/IncludeOS
/usr/bin/clang++-3.8 -std=c++14 -c -Wall -Wextra -O2 -I/home/kuba/IncludeOS_install/api -I/home/kuba/IncludeOS_install/mod/GSL elf_binary.cpp -o elf_binary.o
elf_binary.cpp:79:40: error: non-constant-expression cannot be narrowed from type 'Elf32_Word' (aka 'unsigned int') to 'index_type' (aka 'int') in initializer list [-Wc++11-narrowing]
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
elf_binary.cpp:79:40: note: insert an explicit cast to silence this issue
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
                                       static_cast<index_type>( )
1 error generated.
Makefile:16: recipe for target 'elf_binary.o' failed
make: *** [elf_binary.o] Error 1
/usr/bin/clang++-3.8 -std=c++14 -c -Wall -Wextra -O2 -I/home/kuba/IncludeOS_install/api -I/home/kuba/IncludeOS_install/mod/GSL elf_binary.cpp -o elf_binary.o
elf_binary.cpp:79:40: error: non-constant-expression cannot be narrowed from type 'Elf32_Word' (aka 'unsigned int') to 'index_type' (aka 'int') in initializer list [-Wc++11-narrowing]
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
elf_binary.cpp:79:40: note: insert an explicit cast to silence this issue
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
                                       static_cast<index_type>( )
1 error generated.
Makefile:16: recipe for target 'elf_binary.o' failed
make: *** [elf_binary.o] Error 1
```
(complete installation log is provided at the bottom)

I would be grateful if you could help me with this problem.
If you need any additional information just let me know.

**Sidenote:**
Just to make sure I tried to run `./test.sh` after `./install.sh` (which by the way ends execution with message "Done! Test your installation with ./test.sh" - that is clearly not true since vmbuild compilation failed). `./test.sh` fails with message:
```
~/.installs/IncludeOS/examples/demo_service ~/.installs/IncludeOS

>> Linking service with OS
ld -nostdlib -melf_i386 -N --eh-frame-hdr --script=/home/kuba/IncludeOS_install/linker.ld --defsym _MAX_MEM_MIB_=128 --defsym=__stack_rand_ba=`date +%s` --strip-debug /home/kuba/IncludeOS_install/crt/crtbegin.o /home/kuba/IncludeOS_install/crt/crti.o /home/kuba/IncludeOS_install/multiboot.o service.o .service_name.o /home/kuba/IncludeOS_install/drivers/virtionet.o   /home/kuba/IncludeOS_install/os.a /home/kuba/IncludeOS_install/libcxx/libc++.a /home/kuba/IncludeOS_install/libcxx/libc++abi.a /home/kuba/IncludeOS_install/os.a /home/kuba/IncludeOS_install/newlib/libc.a /home/kuba/IncludeOS_install/newlib/libm.a /home/kuba/IncludeOS_install/libgcc/libgcc.a /home/kuba/IncludeOS_install/crt/crtend.o /home/kuba/IncludeOS_install/crt/crtn.o  -o Demo_Service

>> Adding symbol section
/home/kuba/IncludeOS_install/elf_syms Demo_Service
make: /home/kuba/IncludeOS_install/elf_syms: Command not found
/home/kuba/IncludeOS_install/Makeseed:146: recipe for target 'service' failed
make: *** [service] Error 127

INSTALL FAILED ON COMMAND: make
```

-------
Complete installation log:
```
>>> Installing dependencies (requires sudo):
    Packages: curl make clang-3.8 nasm bridge-utils qemu jq 
Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [94.5 kB]
Ign:2 mirror://mirrors.ubuntu.com/mirrors.txt precise InRelease
Get:3 mirror://mirrors.ubuntu.com/mirrors.txt precise-updates InRelease [55.7 kB]
Hit:4 http://us.archive.ubuntu.com/ubuntu xenial InRelease
Get:5 mirror://mirrors.ubuntu.com/mirrors.txt precise-backports InRelease [55.7 kB]
Get:6 http://us.archive.ubuntu.com/ubuntu xenial-updates InRelease [95.7 kB]
Get:7 mirror://mirrors.ubuntu.com/mirrors.txt precise-security InRelease [55.7 kB]
Get:8 mirror://mirrors.ubuntu.com/mirrors.txt precise Release [49.6 kB]
Hit:9 http://us.archive.ubuntu.com/ubuntu xenial-backports InRelease
Fetched 407 kB in 2s (136 kB/s)
Reading package lists...
W: mirror://mirrors.ubuntu.com/mirrors.txt/dists/precise-updates/InRelease: Signature by key 630239CC130E1A7FD81A27B140976EAF437D05B5 uses weak digest algorithm (SHA1)
W: mirror://mirrors.ubuntu.com/mirrors.txt/dists/precise-backports/InRelease: Signature by key 630239CC130E1A7FD81A27B140976EAF437D05B5 uses weak digest algorithm (SHA1)
W: mirror://mirrors.ubuntu.com/mirrors.txt/dists/precise-security/InRelease: Signature by key 630239CC130E1A7FD81A27B140976EAF437D05B5 uses weak digest algorithm (SHA1)
W: mirror://mirrors.ubuntu.com/mirrors.txt/dists/precise/Release.gpg: Signature by key 630239CC130E1A7FD81A27B140976EAF437D05B5 uses weak digest algorithm (SHA1)
Reading package lists...
Building dependency tree...
Reading state information...
bridge-utils is already the newest version (1.5-9ubuntu1).
make is already the newest version (4.1-6).
jq is already the newest version (1.5+dfsg-1).
nasm is already the newest version (2.11.08-1).
curl is already the newest version (7.47.0-1ubuntu2.2).
clang-3.8 is already the newest version (1:3.8-2ubuntu4).
qemu is already the newest version (1:2.5+dfsg-5ubuntu10.6).
The following packages were automatically installed and are no longer required:
  linux-headers-4.4.0-31 linux-headers-4.4.0-31-generic
  linux-image-4.4.0-31-generic linux-image-extra-4.4.0-31-generic
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 23 not upgraded.
-e 

>>> Calling install_from_bundle.sh script


>>> Getting the ID of the latest release from GitHub
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0 60528    0   453    0     0    355      0  0:02:50  0:00:01  0:02:49   355
100 60528  100 60528    0     0  38996      0  0:00:01  0:00:01 --:--:-- 39000


>>> File to download: https://github.com/hioa-cs/IncludeOS/releases/download/v0.9.0/IncludeOS_install_v0-9-0.tar.gz


>>> IncludeOS tarball exists - extracting to /home/kuba


>>> Fetched tarball - extracting to /home/kuba/IncludeOS_install


>>> Installing submodules
~/.installs/IncludeOS ~/.installs/IncludeOS
~/.installs/IncludeOS


>>> Building IncludeOS
~/.installs/IncludeOS/src ~/.installs/IncludeOS

>>> Built OS-library. Install to '/home/kuba/IncludeOS_install' using 'make install'

>> Installing IncludeOS
cp -r ../api /home/kuba/IncludeOS_install/
mkdir -p /home/kuba/IncludeOS_install/drivers
cp -r drivers/*.o /home/kuba/IncludeOS_install/drivers
mkdir -p /home/kuba/IncludeOS_install/platforms
cp -r platforms/*.o /home/kuba/IncludeOS_install/platforms
cp bootloader /home/kuba/IncludeOS_install
cp boot/multiboot.o /home/kuba/IncludeOS_install
cp os.a /home/kuba/IncludeOS_install
cp libc++abi.a /home/kuba/IncludeOS_install/libcxx
cp crt/crti.o /home/kuba/IncludeOS_install/crt
cp crt/crtn.o /home/kuba/IncludeOS_install/crt
cp linker.ld /home/kuba/IncludeOS_install/
mkdir -p /home/kuba/IncludeOS_install/memdisk
cp memdisk/* /home/kuba/IncludeOS_install/memdisk/
cp seed/service_name.cpp /home/kuba/IncludeOS_install
cp seed/Makefile /home/kuba/IncludeOS_install/Makeseed
cp seed/Makelib /home/kuba/IncludeOS_install/Makelib
cp -r ../seed /home/kuba/IncludeOS_install
mkdir -p /home/kuba/IncludeOS_install/mod/
cp -r ../mod/GSL /home/kuba/IncludeOS_install/mod

Done!
~/.installs/IncludeOS


>>> Compiling the vmbuilder, which makes a bootable vm out of your service
~/.installs/IncludeOS/vmbuild ~/.installs/IncludeOS
/usr/bin/clang++-3.8 -std=c++14 -c -Wall -Wextra -O2 -I/home/kuba/IncludeOS_install/api -I/home/kuba/IncludeOS_install/mod/GSL elf_binary.cpp -o elf_binary.o
elf_binary.cpp:79:40: error: non-constant-expression cannot be narrowed from type 'Elf32_Word' (aka 'unsigned int') to 'index_type' (aka 'int') in initializer list [-Wc++11-narrowing]
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
elf_binary.cpp:79:40: note: insert an explicit cast to silence this issue
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
                                       static_cast<index_type>( )
1 error generated.
Makefile:16: recipe for target 'elf_binary.o' failed
make: *** [elf_binary.o] Error 1
/usr/bin/clang++-3.8 -std=c++14 -c -Wall -Wextra -O2 -I/home/kuba/IncludeOS_install/api -I/home/kuba/IncludeOS_install/mod/GSL elf_binary.cpp -o elf_binary.o
elf_binary.cpp:79:40: error: non-constant-expression cannot be narrowed from type 'Elf32_Word' (aka 'unsigned int') to 'index_type' (aka 'int') in initializer list [-Wc++11-narrowing]
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
elf_binary.cpp:79:40: note: insert an explicit cast to silence this issue
  return {data_.data() + sh.sh_offset, sh.sh_size};
                                       ^~~~~~~~~~
                                       static_cast<index_type>( )
1 error generated.
Makefile:16: recipe for target 'elf_binary.o' failed
make: *** [elf_binary.o] Error 1
~/.installs/IncludeOS


>>> Done! IncludeOS bundle downloaded and installed
-e 

>>> Creating a virtual network, i.e. a bridge. (Requires sudo)
-e 

>>> Done! Test your installation with ./test.sh
```

**Edit no. 1:** Corrections, added note about misguiding information provided after installation finish.

**Comment by JMendyk** [comment]
After a moment of analyzing source code of IncludeOS and figuring out what index_type really is (it's std::ptrdiff_t, but Microsoft's GSL uses "index_type" alias indested) I decided to use suggested by clang static_cast to said std::ptrdiff_t which worked very well and now both `./install.sh` and `./test.sh` execute without any errors. In a moment I'll create pull request with a solution to this problem.

**Comment by ingve** [comment]
I see you reopened this issue. Did the fix in #942 not work for you? Any further information you could provide would be very helpful.

**Comment by JMendyk** [comment]
No, I accidentally closed it and decided to leave it open for a while, since I wasn't sure if project owners want to close them by themselves. I'll close it now since it's not the case.
